File: prev-2024/2264.largest-3-same-digit-number-in-string.js
/*
 * @lc app=leetcode id=2264 lang=javascript
 *
 * [2264] Largest 3-Same-Digit Number in String
 */

// @lc code=start
/**
 * @param {string} num
 * @return {string}
 */
var largestGoodInteger = function (num) {
  let maxInteger = -1;

  for (let i = 1; i < num.length - 1; i++) {
    if (num[i - 1] === num[i] && num[i] === num[i + 1]) {
      maxInteger = Math.max(maxInteger, Number.parseInt(num[i]));
    }
  }

  return maxInteger === -1 ? '' : maxInteger.toString().repeat(3);
};
// @lc code=end

File: src/2025/2264.largest-3-same-digit-number-in-string.cpp
/*
 * @lc app=leetcode id=2264 lang=cpp
 *
 * [2264] Largest 3-Same-Digit Number in String
 */

#include <algorithm>
#include <string>

// @lc code=start
class Solution {
 public:
  std::string largestGoodInteger(const std::string& num) {
    char current_digit = -1;
    int current_digit_substring_length = 0;
    std::string good_integer;

    for (const char digit : num) {
      if (digit != current_digit) {
        current_digit = digit;
        current_digit_substring_length = 1;
      } else {
        current_digit_substring_length++;
        if (current_digit_substring_length == 3 &&
            (good_integer.empty() || current_digit > good_integer[0])) {
          good_integer.resize(3);
          std::ranges::fill(good_integer, current_digit);
        }
      }
    }

    return good_integer;
  }
};
// @lc code=end
